Support/vic.4.2.a.irrigation

hydrotian opened the pull request:
1st PR for vic.4.2.a.irrigaton. Readme file updated.

closes #300

jhamman commented:
Thanks @hydrotian.  This looks pretty good.  

Todo list before we can merge:

- Reviews from:
  - [ ] @tbohn 
  - [ ] @bartnijssen 
  - [ ] Ingrid Tohver (@hydrotian - can you please get in touch wit
  - [ ] @jhamman
- Other items:
  - [x] README update
  - [ ] short sample simulations to demonstrate that this branch is the same as @hydrotian's personal repository

jhamman commented on src/Readme_irrigation.md:
```diff
@@ -0,0 +1,35 @@
+
+Soil file: Same as regular soil file.
+
+
```
suggest formatting this format using markdown syntax.  Current form is hard to read.

jhamman commented on src/full_energy.c:
```diff
@@ -206,24 +237,232 @@ int  full_energy(int                  gridcell,
+	    soil_con->Wcr[lidx] = soil_con->Wcr_orig[lidx];
+	    soil_con->Wpwp[lidx] = soil_con->Wpwp_orig[lidx];
+	  }
+	  } */
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -206,24 +237,232 @@ int  full_energy(int                  gridcell,
+	  veg_var[iveg][band].irrig=0.; //ingjerd added for initialization. hm... not sure about this!
+	  if (options.CROPFRAC  && veg_con[iveg].crop_frac_active)
+	    all_vars_crop->veg_var[veg_con[iveg].crop_frac_idx+1][band].irrig=0.; //ingjerd added for initialization. hm... not sure about this!
+	  //fprintf(stderr,"rec %d sven iveg %d band %d estimating irrigation demand if needed, irrig %f vegvarirrig %f irractive %d\n",rec,iveg,band,irrig,veg_var[iveg][band].irrig,veg_lib[veg_class].irr_active[dmy[rec].month-1]);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -206,24 +237,232 @@ int  full_energy(int                  gridcell,
+	  if(veg_lib[veg_class].irr_active[dmy[rec].month-1]) {  // only apply irrigation to specified crops
+	    //fprintf(stderr,"irr_active in month %d rec %d swq %f airtemp %f \n",dmy[rec].month,rec,snow[iveg][band].swq,atmos->air_temp[NR]);
+	    if(snow[iveg][band].swq<0.001 && atmos->air_temp[NR]>7) { 
+	      //fprintf(stderr,"warm and snow-free\n");
```
suggest removing commented out code x2

Sibada replied:
Will routing be upgrade?


来自 魅族 MX4 Pro

jhamman commented on src/full_energy.c:
```diff
@@ -206,24 +237,232 @@ int  full_energy(int                  gridcell,
+	      
+               if (irr_sm_target > soil_con->max_moist[target_idx])
+                irr_sm_target = soil_con->max_moist[target_idx];
+	       //fprintf(stderr,"full_energy B thresh_idx %d target_idx %d thresh %f target %f\n",thresh_idx,target_idx,irr_sm_thresh,irr_sm_target);
```
suggest removing commented out code x2

jhamman commented on src/full_energy.c:
```diff
@@ -206,24 +237,232 @@ int  full_energy(int                  gridcell,
+                irrig += all_vars_crop->veg_var[veg_con[iveg].crop_frac_idx+1][band].irrig*Cv*soil_con->AreaFract[band];
+		//if(rec==recdummy) 
+		//	fprintf(stderr,"crop_frac_active maxmoist %.2f moist %.2f irrigwishcellaverage %.3f irrigwish(croptile) %.3f Cvirrig %.4f\n",soil_con->max_moist[0],all_vars_crop->cell[veg_con[iveg].crop_frac_idx+1][band].layer[target_idx].moist,irrig,all_vars_crop->veg_var[veg_con[iveg].crop_frac_idx+1][band].irrig,Cv);
+		//fprintf(stderr,"crop_frac_active irrig_est %f irrig %f allvarscropirrig %f moist0 %f\n",irrig_est,irrig,all_vars_crop->veg_var[veg_con[iveg].crop_frac_idx+1][band].irrig,all_vars_crop->cell[1][band].layer[0].moist);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -206,24 +237,232 @@ int  full_energy(int                  gridcell,
+	  }
+	  //if(rec==recdummy)  
+	    //	    fprintf(stderr,"full_energy2 rec %d band %d iveg %d irrig %f moist0 %f \n",
+	    //    rec,band,iveg,irrig,all_vars_crop->cell[1][band].layer[0].moist);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+	 soil_con->Wpwp[lidx] = soil_con->Wpwp_orig[lidx];
+       }
+       }*/
+  
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+          // Limit irrigation to available water
+          if(options.IRRIGATION && veg_lib[veg_class].irr_active[dmy[rec].month-1] && (!options.CROPFRAC || cridx % 2 == 1) && !options.IRR_FREE && atmos->air_temp[NR]>7) { 
+
+	    //fprintf(stderr,"full_energy C2 crop_active rec %d iveg %d cridx %d irrig %f allvarscropirrig %f moist %f\n",rec,iveg,cridx,irrig,all_vars_crop->veg_var[cridx][band].irrig,all_vars_crop->cell[1][band].layer[0].moist); //still cell average
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+	    localextract=1;
+	    if (localextract==1 && irrig>0) {
+	      // extract water from runoff, then baseflow, from each already simulated vegtiles, until reaching irrig.
+	      // forutsetning: irrigated crop is last on vegparam list, and only one veg is irrigated.
```
I'm not familiar with the word "forutsetning"...?

tbohn replied:
must be norwegian

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+	  } //options.irrigation limited
+	  
+	  //	  if(rec==recdummy)  fprintf(stderr,"full_energy I rec %d iveg %d cridx %d veg_class %d irrig_possible (cridx tile) %f irr_run %f\n",
+	  //		     rec,iveg,cridx,veg_class,all_vars_crop->veg_var[cridx][band].irrig,atmos->irr_run[NR]); 
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+				     soil_con, &(all_vars_crop->veg_var[cridx][band]), 
+				     lag_one, sigma_slope, fetch, veg_con[iveg].CanopLayerBnd);
+
+	  //fprintf(stderr,"full_energy D2 rec %d iveg %d cridx %d irrig_wish %f irrig_possible %f temp %f moist %f\n",rec,iveg,cridx,irrig,all_vars_crop->veg_var[cridx][band].irrig,atmos->air_temp[NR],all_vars_crop->cell[1][band].layer[0].moist);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+
+	} /** End non-zero area band **/
+
+	//fprintf(stderr,"\nfull_energy rec %d cridx %d loop crop subtiles, irrig_wish %f moist %f moistirr %f\n",rec,cridx,irrig,all_vars_crop->cell[cridx][band].layer[0].moist,all_vars_crop->cell[1][band].layer[0].moist);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+      } /** end loop over crop subtiles **/
+
+
+      //fprintf(stderr,"\nfull_energy rec %d da var vi ferdig med loop over crop subtiles, irrig_wish %f moist %f\n",rec,irrig,all_vars_crop->cell[1][0].layer[0].moist);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+      for ( band = 0; band < Nbands; band++ ) {
+	if( soil_con->AreaFract[band] > 0 ) {
+
+	  //fprintf(stderr,"full_energy rec %d da begynner vi med backtracking iveg %d fract %f crop_frac %f irrig %f irrig_wish %f\n",rec,iveg,soil_con->AreaFract[band],veg_var[iveg][band].crop_frac,all_vars_crop->veg_var[0][band].irrig,irrig);	  
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+	  veg_var[iveg][band].irrig=all_vars_crop->veg_var[1][band].irrig*veg_var[iveg][band].crop_frac;
+
+	  //fprintf(stderr,"full_energy E ingjerd rec %d iveg %d veg_var.irrig %f allvars.irrig %f\n",rec,iveg,veg_var[iveg][band].irrig,all_vars_crop->veg_var[0][band].irrig);
+	  //fprintf(stderr,"full_energy F ingjerd moist0 %f \n",all_vars_crop->cell[1][band].layer[0].moist);
```
suggest removing commented out code

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+
+	    cell[iveg][band].layer[lidx].moist = (1-veg_var[iveg][band].crop_frac)*all_vars_crop->cell[0][band].layer[lidx].moist + veg_var[iveg][band].crop_frac*all_vars_crop->cell[1][band].layer[lidx].moist;
+
+	    //fprintf(stderr,"\nfull_energy rec %d debugging... irrig_wish %f moist %f\n",rec,irrig,cell[3][0].layer[0].moist);
```
suggest removing commented out code x3

jhamman commented on src/full_energy.c:
```diff
@@ -234,6 +473,547 @@ int  full_energy(int                  gridcell,
+
+      else {
+      
+	//fprintf(stderr,"full_energy here only when irractive for iveg type %d = 0\n",iveg);
```
suggest removing commented out code x2

jhamman commented on src/full_energy.c:
```diff
@@ -417,6 +1195,14 @@ int  full_energy(int                  gridcell,
+	    veg_var[iveg][band].irrig *= (atmos->irr_run[NR]+atmos->irr_with[NR])/irrig; 
+          }
+
+	  //fprintf(stdout,"full_energy n�r og hvor mange ganger kommer vi hit? full_energy G2 rec %d iveg %d irrig_wish %f irrig_possible (mm, incl precip) %f\n",rec,iveg,irrig,veg_var[iveg][band].irrig);
```
suggest removing commented out code

jhamman commented on src/global.h:
```diff
@@ -22,7 +22,7 @@
   2014-May-20 Added ref_veg_vegcover.					TJB
 **********************************************************************/
-char *version = "4.2.1 2014-December-21";
+char *version = "4.2.1 IRR igh tz 2015";
```
Tag is actually 4.2.a.

Also, consider updating to say that this is the `support/VIC.4.2.a.irrigation` branch.

jhamman commented on src/put_data.c:
```diff
@@ -597,6 +599,35 @@ int  put_data(all_vars_struct   *all_vars,
+    if (out_data[OUT_IRRIG].data[0]-out_data[OUT_IRR_EXTRACT].data[0] < out_data[OUT_IRR_RUN].data[0]) {
+      out_data[OUT_IRR_RUN_USED].data[0] = out_data[OUT_IRRIG].data[0]-out_data[OUT_IRR_EXTRACT].data[0];
+      out_data[OUT_IRR_RUN_UNUSED].data[0] = out_data[OUT_IRR_RUN].data[0]-out_data[OUT_IRR_RUN_USED].data[0];
+      //fprintf(stderr,"put_data A1 rec%d out_irrig %f out_irr_extract %f out_irr_run %f\n",rec,out_data[OUT_IRRIG].data[0],out_data[OUT_IRR_EXTRACT].data[0],out_data[OUT_IRR_RUN].data[0]);
```
suggest removing commented out code x2

jhamman commented on src/put_data.c:
```diff
@@ -621,6 +652,14 @@ int  put_data(all_vars_struct   *all_vars,
+  }
+  //fprintf(stderr,"put_data L1 rec %d inflow %f storage %f\n",rec,inflow,out_data[OUT_SOIL_LIQ].data[0]+out_data[OUT_SOIL_LIQ].data[1]+out_data[OUT_SOIL_LIQ].data[2]);
+  //fprintf(stderr,"put_data L2 rec %d inflow %f storage %f\n",rec,inflow,out_data[OUT_SOIL_ICE].data[0]+out_data[OUT_SOIL_ICE].data[1]+out_data[OUT_SOIL_ICE].data[2]);
+  //fprintf(stderr,"put_data L3 rec %d inflow %f swe %f\n",rec,inflow,out_data[OUT_SWE].data[0]);
```
suggest removing commented out code x4

jhamman commented on src/put_data.c:
```diff
@@ -966,6 +1005,19 @@ void collect_wb_terms(cell_data_struct  cell,
+    out_data[OUT_IRRIG].data[0] += veg_var.irrig * AreaFactor;
+    out_data[OUT_IRR_EXTRACT].data[0] += cell.irr_extract * AreaFactor;
+    //fprintf(stderr,"put_data A collect wb terms out_irrig %f irrig %f areafactor %f\n",out_data[OUT_IRRIG].data[0],veg_var.irrig,AreaFactor);
+    //fprintf(stderr,"put_data B collect wb terms out_irr_extract %f cell_irr_extract %f areafactor %f\n",out_data[OUT_IRR_EXTRACT].data[0],cell.irr_extract,AreaFactor);
```
suggest removing commented out code

jhamman commented on src/read_veglib.c:
```diff
@@ -153,6 +153,41 @@ veg_lib_struct *read_veglib(FILE *veglib, int *Ntype)
+      if (options.VEGLIB_IRR) {
+        for (j = 0; j < 12; j++) {
+          fscanf(veglib, "%d", &temp[i].irr_active[j]);
+	  //fprintf(stderr,"class %d month %d irr_active %d\n",i,j,temp[i].irr_active[j]);
```
suggest removing commented out code

jhamman commented on src/read_veglib.c:
```diff
@@ -153,6 +153,41 @@ veg_lib_struct *read_veglib(FILE *veglib, int *Ntype)
+        if (!strcmp(tmpstr,"SAT")) temp[i].irr_sm_target = IRR_SAT;
+        else if (!strcmp(tmpstr,"FC")) temp[i].irr_sm_target = IRR_FC;
+        else if (!strcmp(tmpstr,"CR")) temp[i].irr_sm_target = IRR_CR;
+        //else if (!strcmp(tmpstr,"GSM")) temp[i].irr_sm_target = IRR_GSM; //ingjerd commented out for now
```
suggest removing commented out code x2

jhamman commented on src/surface_fluxes.c:
```diff
@@ -1159,6 +1159,9 @@ int surface_fluxes(char                 overstory,
   (*inflow) = ppt;
+  ppt += veg_var->irrig;
+
+  //fprintf(stdout,"surface_fluxes inflow %f ppt %f veg_varirrig %f\n",(*inflow),ppt,veg_var->irrig);
```
suggest removing commented out code

jhamman replied:
@hydrotian - is there a reason this one is being kept?

hydrotian replied:
@jhamman Oops, I missed this one. Now fixed.

jhamman commented on src/vicNl_def.h:
```diff
@@ -240,6 +240,12 @@
+#define IRR_SAT    0
+#define IRR_FC     1
+#define IRR_CR     2
+//#define IRR_GSM    3 //ingjerd commented out for now
```
suggest removing commented out code

jhamman commented:
@hydrotian - I just read through all this.  Everything looks pretty good to me.  Let's try to get rid of all that commented out code.  

Looking through the diff, I didn't see anything that looks as if it doesn't belong but @tbohn and Ingrid will be able to confirm that.

Lastly, I think we could improve the readme a bit.  A citation to your paper and a bit of formatting would go a long way.

tbohn commented:
Do you really mean Ingrid Tohver?  Perhaps you meant Ingjerd Haddeland?

Her contact info is: iha@nve.no

On Thu, Oct 15, 2015 at 8:05 PM, Joe Hamman <notifications@github.com>
wrote:

> Thanks @hydrotian <https://github.com/hydrotian>. This looks pretty good.
>
> Todo list before we can merge:
>
>    - Reviews from:
>       - @tbohn <https://github.com/tbohn>
>       - @bartnijssen <https://github.com/bartnijssen>
>       - Ingrid Tohver (@hydrotian <https://github.com/hydrotian> - can
>       you please get in touch wit
>       - @jhamman <https://github.com/jhamman>
>    - Other items:
>       - README update
>       - short sample simulations to demonstrate that this branch is the
>       same as @hydrotian <https://github.com/hydrotian>'s personal
>       repository
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/UW-Hydro/VIC/pull/301#issuecomment-148585240>.
>

hydrotian commented:
Thanks Joe, I will follow the to-do list you created. As Ted said, it's Ingjerd Haddeland and her contact is iha@nve.no. I can send her a link to my repo once I finish updating.

jhamman commented:
Sorry about the name confusion, of course it is Ingjerd Haddeland.  I'm back east right now and was working too late last night.  I've updated my checklist accordingly.

hydrotian commented:
@jhamman I just updated the readme file, removed some commented code, and added a bunch of sample input/output files. I also ran this version and compared the results to my previous results and they are the same. By the way, this version doesn't associate with any published paper. I have send the readme file to Ingjerd to see if she has anything to add/change.

jhamman commented on src/Readme_irrigation.md:
```diff
@@ -0,0 +1,34 @@
+# VIC 4.2.a.irrigation
```
@bartnijssen - We haven't fully discussed in the past how we'd like to document these support branches.  Here, @hydrotian has included a README file in the src directory.  I'm wondering if we want to put this documentation somewhere else (either in the header of the root README file or elsewhere).

jhamman commented on src/Readme_irrigation.md:
```diff
@@ -0,0 +1,34 @@
+<br>11 0.0003	 0.30  0.50  0.70  0.50  0
+<br>0.4300  0.8080  1.3730  1.7830  1.2110  0.3660  0.6290  0.9010  0.5010  0.2400  0.2250  0.2960
+<br>79 0.0229	 0.30  0.50  0.70  0.50  1
+<br>1.4460  4.2880  4.5380  2.4280  1.2810  0.3810  1.2810  1.2810  1.2810  1.2810  0.1330  0.6290
```
I think it would look a lot better if you you put this in a code block like this (inside triple ticks):

```
38597 6
7 0.1094	 0.30  0.60  0.70  0.40  0
0.5870  1.0750  1.3500  1.5870  1.7870  1.7120  1.7880  2.1000  2.0370  1.5500  1.0250  0.6620
8 0.2149	 0.30  0.70  0.70  0.30  0
0.2620  0.2870  0.4000  0.4750  0.3000  0.2500  0.4870  0.9250  0.5620  0.2750  0.2250  0.2620
9 0.4242	 0.30  0.70  0.70  0.30  0
0.5500  0.6120  0.7620  0.4500  0.2250  0.3120  0.7250  0.9380  0.4250  0.2000  0.2000  0.3630
10 0.2233	 0.30  0.80  0.70  0.20  0
0.2370  0.2750  0.3880  0.5870  0.8500  1.0500  0.5870  0.4130  0.3130  0.2870  0.2500  0.2250
11 0.0003	 0.30  0.50  0.70  0.50  0
0.4300  0.8080  1.3730  1.7830  1.2110  0.3660  0.6290  0.9010  0.5010  0.2400  0.2250  0.2960
79 0.0229	 0.30  0.50  0.70  0.50  1
1.4460  4.2880  4.5380  2.4280  1.2810  0.3810  1.2810  1.2810  1.2810  1.2810  0.1330  0.6290
```

hydrotian commented:
@jhamman Thanks! Injgerd sent me some notes about changes she made to the code. I will incorporate them into my readme file and update it tonight.

hydrotian commented:
@jhamman @bartnijssen @tbohn I have updated the readme file last week. Could you please review it?

bartnijssen commented:
@hydrotian - you need to merge my PR to your repo, so that the README will update here. I am not going to review the test of it, but will leave that to Ted. For now I unchecked the `README update` box.

tbohn commented:
Just to let you know, I will try to get the review done this week.  Just to
make sure I understand, do you want me to review all of the code changes or
just the README?

Ted

On Mon, Oct 26, 2015 at 3:19 PM, Bart Nijssen <notifications@github.com>
wrote:

> @hydrotian <https://github.com/hydrotian> - you need to merge my PR to
> your repo, so that the README will update here. I am not going to review
> the test of it, but will leave that to Ted. For now I unchecked the README
> update box.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/UW-Hydro/VIC/pull/301#issuecomment-151300800>.
>

bartnijssen commented:
@tbohn: In your case I would like you to look at the code changes to make sure they are consistent with what you gave to Tian.

jhamman commented:
@tbohn - because we had to kluge the history together, we want to make sure we didn't get something in there that didn't belong.

jhamman commented:
@tbohn - we're just waiting on you here.  Shouldn't take long.

bartnijssen commented:
@tbohn: Please review

tbohn commented:
I will try to do this today.

On Fri, Nov 13, 2015 at 9:43 PM, Bart Nijssen <notifications@github.com>
wrote:

> @tbohn <https://github.com/tbohn>: Please review
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/UW-Hydro/VIC/pull/301#issuecomment-156638279>.
>

tbohn commented on src/full_energy.c:
```diff
@@ -206,26 +237,179 @@ int  full_energy(int                  gridcell,
+		  moistlayer=all_vars_crop->cell[0][band].layer[lidx].moist*(1-old_crop_frac)+all_vars_crop->cell[1][band].layer[lidx].moist*old_crop_frac;
+		  moisttotal1+=moistlayer;
+		  all_vars_crop->cell[1][band].layer[lidx].moist = (all_vars_crop->cell[1][band].layer[lidx].moist*old_crop_frac+all_vars_crop->cell[0][band].layer[lidx].moist*(new_crop_frac-old_crop_frac))/new_crop_frac;
+		  all_vars_crop->cell[0][band].layer[lidx].moist = all_vars_crop->cell[0][band].layer[lidx].moist;
```
This line seems redundant - is it just to account for the fact that cell[0] doesn't change?  Perhaps this could be replaced by a comment that cell[0] isn't changing...

tbohn commented on src/full_energy.c:
```diff
@@ -206,26 +237,179 @@ int  full_energy(int                  gridcell,
+            if (cridx % 2 == 0) { //fallow part
+              all_vars_crop->veg_var[cridx][band].vegcover = MIN_VEGCOVER;
+              all_vars_crop->veg_var[cridx][band].albedo = BARE_SOIL_ALBEDO;
+              all_vars_crop->veg_var[cridx][band].LAI = 0;
```
I would suggest adding the following lines here:
all_vars_crop->veg_var[cridx][band].Wdew = 0;
all_vars_crop->veg_var[cridx][band].Wdmax = 0;
all_vars_crop->snow[cridx][band].snow_canopy = 0;

tbohn commented on src/full_energy.c:
```diff
@@ -206,26 +237,179 @@ int  full_energy(int                  gridcell,
+      for(iveg = 0; iveg < Nveg; iveg++){
+        for ( band = 0; band < Nbands; band++ ) {
+	  cell[iveg][band].irr_extract=0; //ingjerd added, for initializing. needed, or maybe not?
+	  veg_class = veg_con[iveg].veg_class; //ingjerd added
```
This could be moved up 2 lines, to just beneath the for(iveg=0...  line.

tbohn commented on src/full_energy.c:
```diff
@@ -206,26 +237,179 @@ int  full_energy(int                  gridcell,
+	  cell[iveg][band].irr_extract=0; //ingjerd added, for initializing. needed, or maybe not?
+	  veg_class = veg_con[iveg].veg_class; //ingjerd added
+	  veg_var[iveg][band].irrig=0.; //ingjerd added for initialization. hm... not sure about this!
+	  if (options.CROPFRAC  && veg_con[iveg].crop_frac_active)
```
I would suggest removing this line.  There is another check on this same condition on line 351 (6 lines below).  Having the if statement on this line (345) prevents irrigation from being computed if CROPFRAC is false or crop_frac_active is false, i.e., denies users the option to run irrigation without dealing with crop fractions.  If you don't think it makes sense to run irrigation without crop fractions turned on, perhaps it would make more sense to get rid of the CROPFRAC option, and hard-code crop_frac_active to true when IRRIGATION is true.

hydrotian replied:
@tbohn I want to confirm about your comment here. Correct me if I'm wrong. To me the if statement (L345) only initialize irrig and it does not prevent computing irrigation if CROPFRAC is false or crop_frac_active is false. So your suggestion is just removing this line (L345) and always initializing irrig?

tbohn replied:
Are our line numbers the same?  My line 345 reads:
if (options.CROPFRAC  && veg_con[iveg].crop_frac_active)
It isn't initializing anything.  It's just preventing the lines below it from running if CROPFRAC isn't true.  The lines below it include all computation of irrigation...

hydrotian replied:
Really? I don't see the braces... Maybe it's the bad indentation here.

tbohn replied:
Ahh, OK, you're right.  Please disregard my earlier comment.

tbohn commented on src/full_energy.c:
```diff
@@ -234,6 +418,500 @@ int  full_energy(int                  gridcell,
+	if( soil_con->AreaFract[band] > 0 ) {
+
+	  //ingjerd added the below sentence, for wb tracking with irrigation and crop_frac included . more is needed?*/
+	  veg_var[iveg][band].irrig=all_vars_crop->veg_var[1][band].irrig*veg_var[iveg][band].crop_frac;
```
This is a very important addition!

tbohn commented on src/initialize_atmos.c:
```diff
@@ -1586,6 +1590,178 @@ void initialize_atmos(atmos_data_struct        *atmos,
+        if (veg_hist[rec][v].crop_frac[0] > 0) {
+          for (j = 0; j < NF; j++) {
+            veg_hist[rec][v].vegcover[j] /= veg_hist[rec][v].crop_frac[0];
+	    veg_hist[rec][v].vegcover[j] =1; //ingjerd added. keep vegcover at 1 for crops
```
What about when crops have only recently been sown or recently emerged from the soil?  Shouldn't vegcover be < 1?  Or is this handled by crop_frac?

In my version, I was setting crop_frac to a (mostly) constant value over the course of a season - it stood for the fraction of the cropland that was planted (perhaps I should have called it planted_frac instead of crop_frac).  So, observed changes in vegcover could be assumed to be due to growth/harvest of plants within the planted fraction.

I suppose this is a moot point. If you are documenting the code that you used, then certainly this should be left as it is.

tbohn replied:
Actually, looking back at full_energy.c, I see that on line 309, you are doing the vegcover /= crop_frac step.  So, really, line 1661 here is incorrect (else line 309 of full_energy.c becomes the second division by crop_frac, which is wrong).  So, I definitely recomend deleting line 1661 and changing the comment on line 1655 to be more descriptive of what you're actually doing - setting vegcover to 1 for crops.

tbohn commented:
OK, I've looked over the changes and made a few comments.  Looks good overall, just a few minor changes I'd recommend (which shouldn't affect the results).  I'm surprised that you all went along with my ugly code - it's not that different from my private version.  But then again, I suppose cleaning up my ugly code would have been a lot of work... sorry about that.

jhamman commented:
Thanks @tbohn - @hydrotian you're up!

hydrotian commented:
Thanks @tbohn and @jhamman! I will do this later today.

hydrotian commented:
@bartnijssen @tbohn @jhamman I have updated the source code and readme file by following Ted's comments. They were all cosmetic changes. I also compared the results to the previous version and made sure the changes did not affect the results. Please review.

jhamman commented:
ping for @hydrotian

jhamman commented:
This seems to be mostly wrapped up.  Anyone apposed to me merging and tagging?

jhamman commented:
Thanks @hydrotian and @tbohn for putting this together.

jhamman commented:
tag is here: https://github.com/UW-Hydro/VIC/releases/tag/VIC.4.2.a.irrigation.01
